design-proposals/klip-57-kafka-headers.md

```diff
@@ -0,0 +1,138 @@
+
+## What is not in scope
+* Writing headers to output topics - we might do this in the future, but that is a separate discussion
+* Advanced functions to help decode byte strings (e.g. functions to deserialize complex serialized data.) - if a user wants to convert something that we don't have a function for, then they can either create their own UDF or use `CREATE FUNCTION` when that becomes available.
+
+## Public APIS
+
+### Syntax
+
+Unlike other pseudocolumns, if a user wants to access headers, they will create a column with the `HEADERS` keyword. This column will be populated with the full list of header keys and values.
+```
+ksql> CREATE STREAM A (my_headers ARRAY<STRUCT<key STRING, value BYTES>> HEADERS);
+```
+The type of a header backed column must be exactly `ARRAY<STRUCT<key STRING, value BYTES>>`. An error will be thrown if that is not the case.
+```
+ksql> CREATE STREAM B (my_headers ARRAY<BYTES> HEADERS);
+Error: Columns specified with the HEADERS keyword must be typed as ARRAY<STRUCT<key STRING, value BYTES>>.
+```
+Users can also mark a column with `HEADER(<key>)`, which will populate the column with the last header that matches the key. The data type must be `BYTES`
+```
+ksql> CREATE STREAM B (value BYTES HEADER(<key>));
+```
+If the key is not present in the header, then the value in the column will be `null`.
+
+Sources created with schema inferencing can have header columns, they would just have to be specified. For example, the statement `CREATE STREAM A (my_headers ARRAY<STRUCT<key STRING, value BYTES>> HEADERS) WITH (kafka_topic='a', value_format='avro');` will create a stream with columns from Schema Registry, as well as one header column.
+
+### New byte-conversion functions
+
+There will also be new byte conversion functions to help decode header data:
+
+* `INT_FROM_BYTES(bytes, byte_order='big', signed=true)` - The input byte array must be exactly 4 bytes long. Signed conversions will by handled by `ByteBuffer`'s `getInt` method. Unsigned conversions will be manually calculated. If the result does not fit into an INTEGER then the method will throw an error.
+* `BIGINT_FROM_BYTES(bytes, byte_order='big', signed=true)` - The input byte array must be exactly 8 bytes long. Signed conversions will by handled by `ByteBuffer`'s `getLong` method. Unsigned conversions will be manually calculated. If the result does not fit into a BIGINT then the method will throw an error.
+* `DOUBLE_FROM_BYTES(bytes, byte_order='big')` - The input byte array must be exactly 8 bytes long. The conversion will by handled by `ByteBuffer`'s `getDouble` method.
+
+### Queries
+
+The headers columns will be usable in any query like any other column, except `INSERT` query will be blocked. However, when persistent queries project header-backed columns, the header values are copied into the output row’s value (or key), not to headers in the output record.
```

Comment:
What does `blocked` mean? Will the `INSERT SELECT` and `INSERT VALUES` fail if trying to insert into a headers column or just blocked/not used?

Reply:
They will fail, added that to the klip